sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -922,6 +923,29 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
   /**
    * Returns the partition names from hive metastore for a given table in a database.
    */
+  override def listPartitionNames(
+      db: String,
+      table: String,
+      partialSpec: Option[TablePartitionSpec] = None): Seq[String] = withClient {
+    val actualPartColNames = getTable(db, table).partitionColumnNames
+    val clientPartitionNames =
+      client.getPartitionNames(db, table, partialSpec.map(lowerCasePartitionSpec))
+
+    if (actualPartColNames.exists(partColName => partColName != partColName.toLowerCase)) {
+      clientPartitionNames.map { partName =>
+        val partSpec = PartitioningUtils.parsePathFragmentAsSeq(partName)
```

Comment:
Sorry, I just realized that isn't exactly what you were asking about.

I think adding coverage for mixed case partition column names is a good thing, but I'll address path (un)escaping as well.